internal/factsengine/gatherers/dir_scan.go

```diff
@@ -0,0 +1,194 @@
+
+type DirScanGatherer struct {
+	fs            afero.Fs
+	userSearcher  UserSearcher
+	groupSearcher GroupSearcher
+}
+
+func NewDirScanGatherer(fs afero.Fs, userSearcher UserSearcher, groupSearcher GroupSearcher) *DirScanGatherer {
+	return &DirScanGatherer{fs: fs, userSearcher: userSearcher, groupSearcher: groupSearcher}
+}
+
+func NewDefaultDirScanGatherer() *DirScanGatherer {
+	cf := CredentialsFetcher{}
+	return NewDirScanGatherer(afero.NewOsFs(), &cf, &cf)
+}
+
+func (d *DirScanGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	log.Infof("Starting %s facts gathering process", DirScanGathererName)
+	facts := []entities.Fact{}
+
+	for _, requestedFact := range factsRequests {
+		if requestedFact.Argument == "" {
+			facts = append(facts, entities.NewFactGatheredWithError(requestedFact, &DirScanMissingArgumentError))
+			continue
+		}
+		scanResult, err := d.extractDirScanDetails(requestedFact.Argument)
+		if err != nil {
+			facts = append(facts, entities.NewFactGatheredWithError(requestedFact, DirScanScanningError.Wrap(err.Error())))
+			continue
+		}
+		factValue, err := mapDirScanResultToFactValue(scanResult)
+		if err != nil {
+			facts = append(facts, entities.NewFactGatheredWithError(requestedFact, DirScanScanningError.Wrap(err.Error())))
+			continue
+		}
+		facts = append(facts, entities.NewFactGatheredWithRequest(requestedFact, factValue))
+	}
+
+	log.Infof("Requested %s facts gathered", DirScanGathererName)
+
+	return facts, nil
+}
+
+func (d *DirScanGatherer) extractDirScanDetails(dirscanPath string) (DirScanResult, error) {
+	result := DirScanResult{}
+
+	matches, err := afero.Glob(d.fs, dirscanPath)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, match := range matches {
+		resultKey := getDirScanResultKeyFromPath(match)
+
+		statInfo, err := d.getStatInfoForPath(match)
+		if err != nil {
+			return nil, err
+		}
+
+		if resultEntry, found := result[resultKey]; found {
+			resultEntry.Files = append(resultEntry.Files, DirScanDetails{
+				Name:  match,
+				Owner: statInfo.Owner,
+				Group: statInfo.Group,
+				Files: nil,
+			})
+			result[resultKey] = resultEntry
+			continue
+		}
+
+		newEntry := DirScanDetails{
+			Owner: statInfo.Owner,
+			Group: statInfo.Group,
+			Files: []DirScanDetails{
+				{
+					Name:  match,
+					Owner: statInfo.Owner,
+					Group: statInfo.Group,
+					Files: nil,
+				},
+			},
+		}
+
+		result[resultKey] = newEntry
+	}
+	return result, nil
+}
+
+func (d *DirScanGatherer) getStatInfoForPath(path string) (*DirScanStatInfo, error) {
+	fi, err := d.fs.Stat(path)
+	if err != nil {
+		return nil, err
+	}
+
+	stat, ok := fi.Sys().(*syscall.Stat_t) //nolint
+	if !ok {
+		return nil, fmt.Errorf("could not extract stat infos for file %s", path)
+	}
+	uid := strconv.Itoa(int(stat.Uid))
+	gid := strconv.Itoa(int(stat.Gid))
+
+	group, err := d.groupSearcher.GetGroupByID(gid)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve group for gigroupd %s", gid)
```

Comment:
typo on `gigroupd`? maybe it should be `gid`

Reply:
done